arangod/Aql/OptimizerRules.cpp

```diff
@@ -60,110 +60,519 @@
+        processModificationNode<arangodb::aql::ReplaceNode>(en);
+        break;
+      }
+
+      default: {
+        // we don't support other node types yet
+        break;
+      }
+    }
+  }
+};
+
+/// WalkerWorker for restrictToSingleShard
+class RestrictToSingleShardChecker final
+    : public arangodb::aql::WalkerWorker<arangodb::aql::ExecutionNode> {
+  arangodb::aql::ExecutionPlan* _plan;
+  CollectionVariableTracker& _tracker;
+  std::unordered_map<arangodb::aql::Variable const*,
+                     std::unordered_set<std::string>>
+      _shardsUsed;
+  std::unordered_map<arangodb::aql::Variable const*,
+                     std::unordered_set<std::string>>
+      _shardsCleared;
+  bool _stop;
+  std::map<arangodb::aql::Collection const*, bool> _unsafe;
+
+ public:
+  explicit RestrictToSingleShardChecker(arangodb::aql::ExecutionPlan* plan,
+                                        CollectionVariableTracker& tracker)
+      : _plan{plan}, _tracker{tracker}, _stop{false} {}
+
+  bool isSafeForOptimization() const {
+    // we have found something in the execution plan that will
+    // render the optimization unsafe
+    return (!_stop && !_plan->getAst()->functionsMayAccessDocuments());
+  }
+
+  std::string getShard(arangodb::aql::Variable const* variable) const {
+    auto const& it = _shardsCleared.find(variable);
+    if (it == _shardsCleared.end()) {
+      return "";
+    }
+
+    auto set = it->second;
+    if (set.size() != 1 || *set.begin() == "all") {
+      return "";
+    }
+
+    return *set.begin();
+  }
+
+  bool isSafeForOptimization(
+      arangodb::aql::Collection const* collection) const {
+    auto it = _unsafe.find(collection);
+    if (it == _unsafe.end()) {
+      return true;
+    }
+    return !it->second;
+  }
+
+  bool isSafeForOptimization(arangodb::aql::Variable const* variable) const {
+    auto it = _shardsCleared.find(variable);
+    if (it == _shardsCleared.end()) {
+      return false;
+    }
+
+    if ((*it).second.size() != 1) {
+      // more than one shard
+      return false;
+    }
+
+    // check for "all" marker
+    auto it2 = (*it).second.find("all");
+    if (it2 != (*it).second.end()) {
+      // "all" included
+      return false;
+    }
+
+    // all good -> safe to optimize
+    return true;
+  }
+
+  bool enterSubquery(arangodb::aql::ExecutionNode*,
+                     arangodb::aql::ExecutionNode*) override final {
+    return true;
+  }
+
+  bool before(arangodb::aql::ExecutionNode* en) override final {
+    using EN = arangodb::aql::ExecutionNode;
+    using arangodb::aql::ExecutionNode;
+
+    switch (en->getType()) {
+      case EN::TRAVERSAL:
+      case EN::SHORTEST_PATH: {
+        _stop = true;
+        return true;  // abort enumerating, we are done already!
+      }
+
+      case EN::FILTER: {
+        auto node = ExecutionNode::castTo<arangodb::aql::FilterNode const*>(en);
+        auto v = node->getVariablesUsedHere();
+        TRI_ASSERT(v.size() == 1);
```

Comment:
where is this enforced ?

Reply:
A FilterNode always has a single input variable, which is normally produced by a CalculationNode.

Reply:
But a FilterNode never has zero or more than one input variables, so the assertion will never fail.